src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs

```diff
@@ -439,47 +439,50 @@ static internal bool IsNonAttributedTypeValidForSerialization(Type type)
-                "System.Collections.Queue",
-                "System.Collections.Stack",
-                "System.Globalization.CultureInfo",
-                "System.Version",
-                "System.Collections.Generic.KeyValuePair`2",
-                "System.Collections.Generic.Queue`1",
-                "System.Collections.Generic.Stack`1",
-                "System.Collections.ObjectModel.ReadOnlyCollection`1",
-                "System.Collections.ObjectModel.ReadOnlyDictionary`2",
-                "System.Tuple`1",
-                "System.Tuple`2",
-                "System.Tuple`3",
-                "System.Tuple`4",
-                "System.Tuple`5",
-                "System.Tuple`6",
-                "System.Tuple`7",
-                "System.Tuple`8",
+        private static readonly Dictionary<string, string[]> s_knownSerializableTypeInfos = new Dictionary<string, string[]> {
+            { "System.Collections.Generic.KeyValuePair`2", Array.Empty<string>() },
+            { "System.Collections.Generic.Queue`1", new[] { "_syncRoot" } },
+            { "System.Collections.Generic.Stack`1", new[] {"_syncRoot" } },
+            { "System.Collections.ObjectModel.ReadOnlyCollection`1", new [] {"_syncRoot" } },
+            { "System.Collections.ObjectModel.ReadOnlyDictionary`2", new [] {"_syncRoot", "_keys","_values" } },
+            { "System.Tuple`1", Array.Empty<string>() },
+            { "System.Tuple`2", Array.Empty<string>() },
+            { "System.Tuple`3", Array.Empty<string>() },
+            { "System.Tuple`4", Array.Empty<string>() },
+            { "System.Tuple`5", Array.Empty<string>() },
+            { "System.Tuple`6", Array.Empty<string>() },
+            { "System.Tuple`7", Array.Empty<string>() },
+            { "System.Tuple`8", Array.Empty<string>() },
+            { "System.Collections.Queue", new[] {"_syncRoot" } },
+            { "System.Collections.Stack", new[] {"_syncRoot" } },
+            { "System.Globalization.CultureInfo", Array.Empty<string>() },
+            { "System.Version", Array.Empty<string>() },
         };
 
+        private static string GetGeneralTypeName(Type type)
+        {
+            TypeInfo typeInfo = type.GetTypeInfo();
+            return type.GetTypeInfo().IsGenericType && !typeInfo.IsGenericParameter
+                ? typeInfo.GetGenericTypeDefinition().FullName
+                : type.FullName;
+        }
+
         internal static bool IsKnownSerializableType(Type type)
         {
             // Applies to known types that DCS understands how to serialize/deserialize
             //
+            string typeFullName = GetGeneralTypeName(type);
 
-            // Ajdust for generic type
-            if (type.GetTypeInfo().IsGenericType && !type.GetTypeInfo().IsGenericTypeDefinition)
-            {
-                type = type.GetGenericTypeDefinition();
-            }
+            return s_knownSerializableTypeInfos.Keys.Contains(typeFullName)
```

Comment:
`return s_knownSerializableTypeInfos.ContainsKey(typeFullName)` to avoid the `KeyCollection` allocation when accessing the `Keys` property.